### PR TITLE
Redesign sawtooth adapter to make it easy to add new use uses

### PR DESCRIPTION
**Pull request by Ram-srini**
1. Generalize the sawtooth adapter
2. Move use case specific code like batch creation and adddress creation to application adapter
3. Each application has to implement batch builder interface.

Signed-off-by: Ramakrishna Srinivasamurthy <ramakrishna.srinivasamurthy@intel.com>

### CODE REVIEW

**Comment by Ram-srini** [comment]
@haojun #32 I accidentally clicked on close and comment.

**Comment by Ram-srini** [comment]
Idea is to not change the sawtooth adapter for every new use case. Building payload and address calculation is use case specific. Use case which defines iniit(), run() and end(), is also meant to generic. Because we it is going consumed by other hyper ledger.

File: src/sawtooth/Application/BatchBuilder.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+'use strict'
+
+class BatchBuilder {
+    constructor() {
+    }
+
+    buildBatch(args) {
+        throw new Error('buildiBatch is not implemented for this application!!');
```
Comment (dcmiddle) [inline src/sawtooth/Application/BatchBuilder.js]: typo: buildiBatch / buildBatch

**Comment by Ram-srini** [comment]
@dcmiddle thanks Dan for pointing out typo error. I have corrected the typo error.

**Comment by Ram-srini** [comment]
@feihujiang Please review changes.

File: src/sawtooth/Application/BatchBuilderFactory.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+'use strict'
+
+var SimpleBatchBuilder = require('./SimpleBatchBuilder.js')
+var SmallBankBatchBuilder = require('./SmallBankBatchBuilder.js')
+
```
Comment (feihujiang) [inline src/sawtooth/Application/BatchBuilderFactory.js]: Where is the ./SmallBankBatchBuilder.js?

**Comment by Ram-srini** [comment]
@feihujiang I have removed those changes.

File: src/sawtooth/Application/BatchBuilder.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+'use strict'
```
Comment (feihujiang) [inline src/sawtooth/Application/BatchBuilder.js]: Should have license text.

File: src/sawtooth/Application/BatchBuilderFactory.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+'use strict'
```
Comment (feihujiang) [inline src/sawtooth/Application/BatchBuilderFactory.js]: Should have license text

File: src/sawtooth/Application/SimpleBatchBuilder.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
+'use strict'
```
Comment (feihujiang) [inline src/sawtooth/Application/SimpleBatchBuilder.js]: Should have license text.

File: src/sawtooth/Application/BatchBuilder.js (hunk 3 above)
Comment (Ram-srini) [reply src/sawtooth/Application/BatchBuilder.js]: Done.

File: src/sawtooth/Application/BatchBuilderFactory.js (hunk 4 above)
Comment (Ram-srini) [reply src/sawtooth/Application/BatchBuilderFactory.js]: Done

File: src/sawtooth/Application/SimpleBatchBuilder.js (hunk 5 above)
Comment (Ram-srini) [reply src/sawtooth/Application/SimpleBatchBuilder.js]: Done

**Pull request merged**